File: App/Resources/AlgorithmDetails/improvedblockselectionsort/improvedblockselectionsort.py
def block_root(n):
    i = 1
    while i * i < n:
        i *= 2
    return i


def multi_swap(array, a, b, length):
    for i in range(length):
        array[a + i], array[b + i] = array[b + i], array[a + i]


def rotate(array, a, m, b):
    l = m - a
    r = b - m
    while l > 0 and r > 0:
        if r < l:
            multi_swap(array, m - r, m, r)
            b -= r
            m -= r
            l -= r
        else:
            multi_swap(array, a, m, l)
            a += l
            m += l
            r -= l


def select_range(array, start, end, b_len):
    min_index = start
    a = start + b_len
    while a < end:
        if (
            array[a] < array[min_index]
            or array[a] == array[min_index]
            and array[a + b_len - 1] < array[min_index + b_len - 1]
        ):
            min_index = a
        a += b_len
    return min_index


def block_select(array, a, m, b, b_len):
    k = a
    j = m
    while k < m and array[k] <= array[m]:
        k += b_len
    if k == m:
        return

    i = m
    multi_swap(array, k, j, b_len)
    k += b_len
    j += b_len

    while k < j < b:
        if array[i] <= array[j]:
            if k != i:
                multi_swap(array, k, i, b_len)
            k += b_len
            i = select_range(array, max(m, k), j, b_len)
        else:
            if i == k:
                i = j
            if k != j:
                multi_swap(array, k, j, b_len)
            k += b_len
            j += b_len

    while k < j:
        i = select_range(array, k, b, b_len)
        if k != i:
            multi_swap(array, k, i, b_len)
        k += b_len


def in_place_merge(array, a, m, b):
    i = a
    j = m
    while i < j < b:
        if array[i] > array[j]:
            k = j + 1
            while k < b and array[i] > array[k]:
                k += 1
            rotate(array, i, j, k)
            i += k - j
            j = k
        else:
            i += 1
    return i


def in_place_merge_bw(array, a, m, b):
    i = m - 1
    j = b - 1
    while j > i >= a:
        if array[i] > array[j]:
            k = i - 1
            while k >= a and array[k] > array[j]:
                k -= 1
            rotate(array, k + 1, i + 1, j + 1)
            j -= i - k
            i = k
        else:
            j -= 1


def sort(array):
    n = len(array)
    if n <= 1:
        return
    j = 1
    while j < n:
        b_len = block_root(j)
        run_length = j
        b = n - n % b_len

        while run_length > 16:
            i = 0
            while i + j < b:
                k = i
                while k + run_length < min(i + 2 * j, b):
                    block_select(
                        array, k, k + run_length, min(k + 2 * run_length, b), b_len
                    )
                    k += run_length
                i += 2 * j
            run_length = b_len
            b_len = block_root(b_len)

        i = 0
        while i + j < b:
            k = i
            f = i
            while k + run_length < min(i + 2 * j, b):
                f = in_place_merge(array, f, k + run_length, min(k + 2 * run_length, b))
                k += run_length
            i += 2 * j

        in_place_merge_bw(array, n - n % (2 * j), b, n)
        j *= 2


if __name__ == "__main__":
    array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56]
    sort(array)
    print(array)

File: App/Resources/AlgorithmDetails/improvedblockselectionsort/improvedblockselectionsort.cpp
#include <algorithm>
#include <cstdio>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

int blockRoot(int n) {
  int i = 1;
  while (i * i < n) {
    i *= 2;
  }
  return i;
}

void multiSwap(int arr[], int a, int b, int len) {
  for (int i = 0; i < len; i++) {
    std::swap(arr[a + i], arr[b + i]);
  }
}

void rotate(int arr[], int a, int m, int b) {
  int l = m - a;
  int r = b - m;
  while (l > 0 && r > 0) {
    if (r < l) {
      multiSwap(arr, m - r, m, r);
      b -= r;
      m -= r;
      l -= r;
    } else {
      multiSwap(arr, a, m, l);
      a += l;
      m += l;
      r -= l;
    }
  }
}

int selectRange(int arr[], int start, int end, int bLen) {
  int minIndex = start;
  int a = start + bLen;
  while (a < end) {
    if (arr[a] < arr[minIndex]) {
      minIndex = a;
    } else if (arr[a] == arr[minIndex] &&
               arr[a + bLen - 1] < arr[minIndex + bLen - 1]) {
      minIndex = a;
    }
    a += bLen;
  }
  return minIndex;
}

void blockSelect(int arr[], int a, int m, int b, int bLen) {
  int k = a;
  int j = m;
  while (k < m && arr[k] <= arr[m]) {
    k += bLen;
  }
  if (k == m) {
    return;
  }

  int i = m;
  multiSwap(arr, k, j, bLen);
  k += bLen;
  j += bLen;

  while (k < j && j < b) {
    if (arr[i] <= arr[j]) {
      if (k != i) {
        multiSwap(arr, k, i, bLen);
      }
      k += bLen;
      i = selectRange(arr, std::max(m, k), j, bLen);
    } else {
      if (i == k) {
        i = j;
      }
      if (k != j) {
        multiSwap(arr, k, j, bLen);
      }
      k += bLen;
      j += bLen;
    }
  }

  while (k < j) {
    i = selectRange(arr, k, b, bLen);
    if (k != i) {
      multiSwap(arr, k, i, bLen);
    }
    k += bLen;
  }
}

int inPlaceMerge(int arr[], int a, int m, int b) {
  int i = a;
  int j = m;
  while (i < j && j < b) {
    if (arr[i] > arr[j]) {
      int k = j + 1;
      while (k < b && arr[i] > arr[k]) {
        k++;
      }
      rotate(arr, i, j, k);
      i += k - j;
      j = k;
    } else {
      i++;
    }
  }
  return i;
}

void inPlaceMergeBW(int arr[], int a, int m, int b) {
  int i = m - 1;
  int j = b - 1;
  while (j > i && i >= a) {
    if (arr[i] > arr[j]) {
      int k = i - 1;
      while (k >= a && arr[k] > arr[j]) {
        k--;
      }
      rotate(arr, k + 1, i + 1, j + 1);
      j -= i - k;
      i = k;
    } else {
      j--;
    }
  }
}

void sort(int arr[], int n) {
  if (n <= 1) {
    return;
  }
  int j = 1;
  while (j < n) {
    int bLen = blockRoot(j);
    int runLength = j;
    int b = n - n % bLen;

    while (runLength > 16) {
      int i = 0;
      while (i + j < b) {
        int k = i;
        while (k + runLength < std::min(i + 2 * j, b)) {
          blockSelect(arr, k, k + runLength, std::min(k + 2 * runLength, b),
                      bLen);
          k += runLength;
        }
        i += 2 * j;
      }
      runLength = bLen;
      bLen = blockRoot(bLen);
    }

    int i = 0;
    while (i + j < b) {
      int k = i;
      int f = i;
      while (k + runLength < std::min(i + 2 * j, b)) {
        f = inPlaceMerge(arr, f, k + runLength, std::min(k + 2 * runLength, b));
        k += runLength;
      }
      i += 2 * j;
    }

    inPlaceMergeBW(arr, n - n % (2 * j), b, n);
    j *= 2;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
